nearswap/src/twap.rs

```diff
@@ -0,0 +1,601 @@
+                end = mid;
+            } else {
+                start = mid + 1;
+            }
+        }
+
+        if self.pivoted && start == 0 {
+            let res = start;
+            start = self.current_idx + 1;
+            end = max_length;
+
+            while start < end {
+                mid = (start + end) / 2;
+                if block_timestamp <= self.observations.get(mid).unwrap().block_timestamp {
+                    end = mid;
+                } else {
+                    start = mid + 1;
+                }
+            }
+            if start >= max_length {
+                start = res;
+            }
+        }
+
+        return start;
+    }
+
+    // function which will calculate mean using Mean enum
+    pub fn calculate_mean(
+        &self,
+        time: Mean,
+        max_length: u64,
+    ) -> (u128, u128) {
+        let time_diff: u64 = match time {
+            Mean::M_1MIN => T_1MIN, // 1 minute in nanoseconds
+            Mean::M_5MIN => T_5MIN, // 5 minute in nanoseconds
+            Mean::M_1H => T_1H,
+            Mean::M_12H => T_12H,
+            _ => 0
+        };
+        let req_timestamp;
+        if self.observations.get(self.current_idx).unwrap().block_timestamp >= time_diff {
+            req_timestamp = self.observations.get(self.current_idx).unwrap().block_timestamp - time_diff;
+        } else {
+            req_timestamp = self.observations.get(self.current_idx).unwrap().block_timestamp;
+        }
+         
+        let left_index = self.binary_search(max_length, req_timestamp);
+        let current_o = &self.observations.get(self.current_idx).unwrap();
+        
+        let total_observe;
+        let price1cumu;
+        let price2cumu;
+        if left_index == 0 {
+            total_observe = current_o.num_of_observations;
+            price1cumu = current_o.price1_cumulative;
+            price2cumu = current_o.price2_cumulative;
+        } else {
+            let prev_o = &self.observations.get(left_index - 1).unwrap();
+            total_observe = current_o.num_of_observations
+                                    - prev_o.num_of_observations;
+            price1cumu = current_o.price1_cumulative
+                                    - prev_o.price1_cumulative;
+            price2cumu = current_o.price2_cumulative
+                                    - prev_o.price2_cumulative;
+        }
+        let mean1 = price1cumu / total_observe;
+        let mean2 = price2cumu / total_observe;
+        return (mean1, mean2);
+    }
+
+    pub(crate) fn log_observation(&mut self, timestamp: u64, price1: u128, price2: u128) {
+        // update current index
+        if self.observations.len() == 0 {
+            self.current_idx = self.initialize(timestamp, price1, price2);
+        } else {
+            self.current_idx = self.write(
+                timestamp,
+                price1,
+                price2,
+                MAX_LENGTH
```

Comment:
We shouldn't pass `MAX_LENGTH` as a parameter. I guess you have it only to make it easier to write tests. If this a the case, then you can use a class static variable for that - and modify it in tests. Other solution is to put in in the TWAP structure and initialize it in the constructor (constructor can use it as a parameter) - this is less optimal, because each object will consume a bit more space. But it's not bad - it's only one small variable - the array has much more storage.

Reply:
ok updated